Improve size-dependent phytoplankton growth code in BLING

averdy opened the pull request:
## What changes does this PR introduce?
New feature: size-dependent growth limitation coefficients are now run-time parameters;
Tidy up code by merging SIZE_DEP_LIM and SIZE_NUT_LIM options;
Bug fix: add ifdef statement in bling_bio_nitrogen for the code to compile when option SIZE_DEP_LIM is undefined.


## What is the current behaviour? 
Size-dependent growth limitation coefficients are hard-coded in bling_bio_nitrogen.F;
SIZE_DEP_LIM (determines growth limitation coefficient values) can be defined while SIZE_NUT_LIM (determines if phyto growth uses CaCO3) is undefined, or vice versa;
Code doesn't compile when SIZE_DEP_LIM is undefined.


## What is the new behaviour 
Size-dependent phytoplankton parameters can be specified in data.bling;
Size dependence affects both growth rate and CaCO3 update, not just one or the other;
Code compiles when size dependence is not defined. 


## Does this PR introduce a breaking change? 
No


## Other information:


## Suggested addition to `tag-index`
pkg/bling: size-dependent phytoplankton growth code cleaned up and related parameters can now be specified in data.bling

kunalmadkaiker commented on pkg/bling/bling_bio_nitrogen.F:
```diff
@@ -888,15 +841,13 @@ SUBROUTINE BLING_BIO_NITROGEN(
 C  with the findings of Jin et al. (GBC,2006).
 
-CAV(
 #ifdef SIZE_NUT_LIM
```
Here SIZE_NUT_LIM should be replaced by SIZE_DEP_LIM

mmazloff commented:
LGTM!

mjlosch commented on pkg/bling/BLING_OPTIONS.h:
```diff
@@ -30,8 +30,8 @@ C nutrient limitations to calculate maximum phyto growth rate
-C nutrient/light limitations. Parameters implemented have yet to be tuned.
-#undef SIZE_NUT_LIM
+C nutrient/light limitations. Parameters implemented have yet to be tuned
+#define SIZE_DEP_LIM
```
Is it a good idea to have a new default here? I would undef this by default, as in the `verifcation/global_oce_biogeo_bling/code` it defined in local copy anyway, but in `code_ad` it isn't (and I am surprised that the value of SIZE_DEP_LIM has no effect on the results).

averdy replied:
Sure, we can have it undefined by default. The reason it doesn't change the results is that the forward verification setup doesn't have phytoplankton growing at all - this is due to #define USE_QSW by default. That issue is addressed in PR #841 ... but since we're probably not going forward with the exf thing, maybe the default can be changed here to #undef USE_QSW?

mjlosch reviewed (approved):
Looks good in general. Please note my two comments

mmazloff commented:
LGTM.

mjlosch commented on pkg/bling/BLING_VARS.h:
```diff
@@ -274,6 +274,14 @@ C          satellite-equivalent chlorophyll diagnostic (and cost)
      &                     gamma_DON,
      &                     k_Fe_diaz,
      &                     k_NO3,
+#ifdef SIZE_DEP_LIM
+     &                     k_NO3_sm,
+     &                     k_NO3_lg,
+     &                     k_PO4_sm,
+     &                     k_PO4_lg,
+     &                     k_Fe_sm,
+     &                     k_Fe_lg,
+#endif
```
I think that we normally have namelist parameters defined independent of the CPP-flag:
```suggestion
     &                     k_NO3_sm,
     &                     k_NO3_lg,
     &                     k_PO4_sm,
     &                     k_PO4_lg,
     &                     k_Fe_sm,
     &                     k_Fe_lg,
```

averdy replied:
@mjlosch Interesting! I modeled this after pkg/dic, where some namelist parameters are only defined #ifdef DIC_BIOTIC or #ifdef DIC_CALCITE_SAT.

We have a few CPP flags in BLING_VARS.h and bling_readparms.F, e.g. #ifdef USE_SIBLING and #ifdef USE_BLING_V1. Should those be removed too?

mjlosch replied:
So this seems to be a BGC-specific method? (Kidding). 

I think it makes a lot of sense to always have the namelist parameters defined, just because the namelist error messages are often very unintuitive, if a parameter is not recognised. Downside: You don't notice that you have accidently forgotten to set a CPP flag, read the namelist and it has no effect. Normally we have some kind of check for this (e.g. check for unset parameters), but this would be very difficult, when you have defaults as in this case. 

In the end, it's up to you, as you and your group are probably the main users of this pkg.

mjlosch commented on pkg/bling/BLING_VARS.h:
```diff
@@ -376,6 +384,14 @@ C          satellite-equivalent chlorophyll diagnostic (and cost)
       _RL gamma_DON
       _RL k_Fe_diaz
       _RL k_NO3
+#ifdef SIZE_DEP_LIM
+      _RL k_NO3_sm
+      _RL k_NO3_lg
+      _RL k_PO4_sm
+      _RL k_PO4_lg
+      _RL k_Fe_sm
+      _RL k_Fe_lg
+#endif
```
```suggestion
      _RL k_NO3_sm
      _RL k_NO3_lg
      _RL k_PO4_sm
      _RL k_PO4_lg
      _RL k_Fe_sm
      _RL k_Fe_lg
```

mjlosch commented on pkg/bling/bling_readparms.F:
```diff
@@ -107,6 +107,14 @@ SUBROUTINE BLING_READPARMS( myThid )
      &                     k_NO3,
      &                     k_PtoN,
      &                     k_FetoN,
+# ifdef SIZE_DEP_LIM
+     &                     k_NO3_sm,
+     &                     k_NO3_lg,
+     &                     k_PO4_sm,
+     &                     k_PO4_lg,
+     &                     k_Fe_sm,
+     &                     k_Fe_lg,
+# endif
```
```suggestion
     &                     k_NO3_sm,
     &                     k_NO3_lg,
     &                     k_PO4_sm,
     &                     k_PO4_lg,
     &                     k_Fe_sm,
     &                     k_Fe_lg,
```

mjlosch commented on pkg/bling/bling_readparms.F:
```diff
@@ -341,6 +349,14 @@ SUBROUTINE BLING_READPARMS( myThid )
       k_Fe_diaz            = 7. _d -10 / permil
       k_NO3                = 2. _d -6 / permil
       k_PO4                = 1. _d -8 / permil
+# ifdef SIZE_DEP_LIM
+      k_NO3_sm             = 2.0 _d -6 / permil
+      k_NO3_lg             = 6.0 _d -6 / permil
+      k_PO4_sm             = 1.0 _d -8 / permil
+      k_PO4_lg             = 3.0 _d -8 / permil
+      k_Fe_sm              = 1.6 _d -10 / permil
+      k_Fe_lg              = 4.8 _d -10 / permil
+# endif
```
```suggestion
      k_NO3_sm             = 2.0 _d -6 / permil
      k_NO3_lg             = 6.0 _d -6 / permil
      k_PO4_sm             = 1.0 _d -8 / permil
      k_PO4_lg             = 3.0 _d -8 / permil
      k_Fe_sm              = 1.6 _d -10 / permil
      k_Fe_lg              = 4.8 _d -10 / permil
```

mjlosch reviewed (commented):
It's probably more convenient to have the namelist parameters always defined, in line with our general policy (avoid weird error messages when `SIZE_DEP_LIM` is not defined, but the namelist has not yet changed)

averdy commented:
@mjlosch I get your point, and will try to follow that advice moving forward. However, since we have so many namelist parameters dependent on CPP-flags already, I'd like to leave the bling code as is for now.

averdy commented on pkg/bling/BLING_OPTIONS.h:
```diff
@@ -87,13 +88,5 @@ C  coeffs are used).
-C-- Undocumented Options:
-#undef MULT_NUT_LIM
-#undef USE_ATMOSCO2
-#undef USE_EXFCO2
```
Done, thanks!

jm-c commented:
@averdy I have few comments regarding changes to `pkg/bling/BLING_OPTIONS.h`:
1. I run the forward test exp. `global_oce_biogeo_bling`, and since the (default) setting of options is used in the test, the results are different enough to cause this test to fail. I re-run it with all other modified src code but keeping the old `BLING_OPTIONS,h` and can reproduce the old results. I don't know which option setting update is causing these differences but we need to decide if we want to keep the same results (and update the default BLING_OPTIONS.h) or do an update of the reference output.
2. I think I did add, some time ago, the list of few "Undocumented Options" at the bottom of `BLING_OPTIONS.h`, because it's useful to know where a user can turn on/off an option that shows up somewhere in the src code. With this PR, "USE_ATMOSCO2" is gone and "MULT_NUT_LIM" is properly documented. But "USE_EXFCO2" can still be found in src code so I would prefer to keep it there, or remove it completely from any src code.
3. a minor edit: since "MULT_NUT_LIM" is only relevant with USE_BLING_V1, it could be moved just after "#undef USE_BLING_V1", before "Options for BLING+Nitrogen code:".

averdy commented:
@jm-c Thank you for the comments!
1. The change comes from undefining USE_QSW. Since the verification experiment doesn't use EXF, it's silly to have that option defined because it leads to total darkness and no phytoplankton growth. Eventually I would like to change the verification experiment to one that does use EXF. We can keep USE_QSW defined until then, if it's easier to not to have to change the results. Let me know what you prefer!
2. I see that USE_EXFCO2 is in pkg/exf/exf_monitor.F, where the atmospheric pCO2 value is printed out. I imagine that will be handled differently after PR #876 ... Would it be ok to remove lines 213-218 from exf_monitor.F?
3. Moved as suggested.

jm-c commented:
@averdy I think there is some confusion regarding my point (3): I meant "MULT_NUT_LIM" and not "MIN_NUT_LIM" (which is used outside of BLING_v1 code).

And regarding option "USE_EXFCO2", I could suggest to replace it with "#ifdef ALLOW_BLING" in `pkg/exf_monitor.F` to make minimal changes (and it will be sorted out later in PR #876).

Regarding "USE_QSW", I think I understand your point, and it's better to have it "#undef" (as a default and in this fwd test exp.), plus it's already "#undef" in the AD test exp. So I could generate a new ref. output for the fwd test and we need to mention this fix/update in this PR description. Do you agree ?

averdy commented:
@jm-c I did get myself confused! The two flags are basically opposites. So I replaced MULT_NUT_LIM with MIN_NUT_LIM in bling_bio.F and removed the flag from BLING_OPTIONS.h. It will only cause a change for setups that use BLING_V1 and #define MULT_NUT_LIM, and I don't think anyone is using V1... Do you think that's ok? 

I followed your suggestion for exf_monitor.F

I agree with having #undef USE_QSW and updating the results. 

Thanks!

jm-c commented:
@averdy OK, I like your solution. I will tweak a bit few files and than get a new ref. output.

jm-c commented:
Things look good here. I've checked that both the TAF and Tapenade AD tests are not affected. Could merge this PR soon, after @averdy you agree.

averdy commented:
@jm-c Agreed!
